File: tests/unit/consumption.test.ts
// XTRITIUM §5.2 — the full-tank algorithm, checked against its own wording.
//
// This is the first figure TRITIUM computes rather than stores, which is also
// the first figure that can be quietly wrong for a year. So each clause of
// §5.2 is its own test, and the arithmetic is checked against sums done by
// hand rather than against the implementation's own output.

import { describe, expect, it } from 'vitest'
import {
  consumptionAt,
  consumptionById,
  consumptionPoints,
  entryTotal,
  sortByOdometer
} from '../../src/shared/consumption.js'
import type { FuelEntry } from '../../src/shared/records.js'
import { toPump } from '../../src/shared/scaled.js'

function fill(
  id: string,
  odometer: number,
  litres: number,
  full: boolean,
  date = '2026-01-01'
): FuelEntry {
  return {
    id,
    date,
    odometer_km: odometer,
    litres: toPump(litres),
    price_per_litre: toPump(73.38),
    full_tank: full,
    fuel_type: 'Kurşunsuz 95'
  }
}

describe('a consumption data point', () => {
  it('exists only at a full tank that has an earlier full tank before it', () => {
    const points = consumptionPoints([
      fill('f-0001', 19_000, 40, true),
      fill('f-0002', 19_500, 30, true)
    ])

    // 30 l over 500 km — 6,00 l/100km, by hand.
    expect(points).toHaveLength(1)
    expect(points[0]?.id).toBe('f-0002')
    expect(points[0]?.distance_km).toBe(500)
    expect(points[0]?.litres).toBe(toPump(30))
    expect(points[0]?.l100km).toBe(6_000)
  })

  it('does not exist at the first-ever entry', () => {
    expect(consumptionPoints([fill('f-0001', 19_000, 40, true)])).toEqual([])
  })

  it('does not exist at a partial entry', () => {
    const points = consumptionPoints([
      fill('f-0001', 19_000, 40, true),
      fill('f-0002', 19_500, 30, true),
      fill('f-0003', 19_800, 20, false)
    ])

    // The trailing partial adds a third entry and no third figure.
    expect(points.map((point) => point.id)).toEqual(['f-0002'])
  })

  it('never exists when nothing was ever filled to the top', () => {
    expect(
      consumptionPoints([
        fill('f-0001', 19_000, 40, false),
        fill('f-0002', 19_500, 30, false),
        fill('f-0003', 20_000, 30, false)
      ])
    ).toEqual([])
  })
})

describe('the litres a point counts', () => {
  it('adds every partial fill in between to the full entry that closes the interval', () => {
    const points = consumptionPoints([
      fill('f-0001', 19_000, 40, true),
      fill('f-0002', 19_200, 10, false),
      fill('f-0003', 19_500, 30, true)
    ])

    // 30 + 10 = 40 l over 500 km — 8,00 l/100km.
    expect(points).toHaveLength(1)
    expect(points[0]?.litres).toBe(toPump(40))
    expect(points[0]?.l100km).toBe(8_000)
  })

  it('counts two partials in the same interval', () => {
    const points = consumptionPoints([
      fill('f-0001', 10_000, 50, true),
      fill('f-0002', 10_150, 12.5, false),
      fill('f-0003', 10_300, 12.5, false),
      fill('f-0004', 10_600, 35, true)
    ])

    // 35 + 12,5 + 12,5 = 60 l over 600 km — 10,00 l/100km.
    expect(points).toHaveLength(1)
    expect(points[0]?.litres).toBe(toPump(60))
    expect(points[0]?.l100km).toBe(10_000)
  })

  it('drops a partial that comes before the first full tank', () => {
    // It belongs to no interval: there is no earlier full tank to measure from,
    // so its litres are not owed to the first interval that does exist.
    const points = consumptionPoints([
      fill('f-0001', 19_000, 15, false),
      fill('f-0002', 19_100, 40, true),
      fill('f-0003', 19_600, 30, true)
    ])

    expect(points).toHaveLength(1)
    expect(points[0]?.litres).toBe(toPump(30))
    expect(points[0]?.distance_km).toBe(500)
  })

  it('starts a fresh interval after each full tank', () => {
    const points = consumptionPoints([
      fill('f-0001', 10_000, 50, true),
      fill('f-0002', 10_400, 20, true),
      fill('f-0003', 10_600, 10, false),
      fill('f-0004', 10_900, 35, true)
    ])

    expect(points.map((point) => point.litres)).toEqual([toPump(20), toPump(45)])
    expect(points.map((point) => point.distance_km)).toEqual([400, 500])
  })
})

describe('the order the entries arrive in', () => {
  it('is the odometer, not the file', () => {
    // §5.2 opens by sorting on the odometer. A fill-up entered late still
    // happened at the reading it carries.
    const sorted = sortByOdometer([
      fill('f-0003', 19_500, 30, true),
      fill('f-0001', 19_000, 40, true),
      fill('f-0002', 19_200, 10, false)
    ])

    expect(sorted.map((entry) => entry.odometer_km)).toEqual([19_000, 19_200, 19_500])
  })

  it('produces the same figures however the list is shuffled', () => {
    const entries = [
      fill('f-0001', 19_000, 40, true),
      fill('f-0002', 19_200, 10, false),
      fill('f-0003', 19_500, 30, true)
    ]

    const forwards = consumptionPoints(entries)
    const backwards = consumptionPoints([...entries].reverse())
    expect(backwards).toEqual(forwards)
  })

  it('places an entry typed with a lower odometer by its reading', () => {
    // §5.1 accepts a backwards odometer — the maker's word is final. Sorting on
    // the reading is what that acceptance means downstream: the entry lands
    // where its number says, and the intervals close around it.
    const points = consumptionPoints([
      fill('f-0001', 19_000, 40, true),
      fill('f-0002', 19_800, 30, true),
      fill('f-0003', 19_400, 20, true)
    ])

    expect(points.map((point) => point.odometer_km)).toEqual([19_400, 19_800])
    expect(points.map((point) => point.distance_km)).toEqual([400, 400])
  })

  it('says nothing about two entries at the same reading', () => {
    // Zero distance measures nothing. The interval is skipped; the second entry
    // still becomes the mark the next one is measured from.
    const points = consumptionPoints([
      fill('f-0001', 19_000, 40, true),
      fill('f-0002', 19_000, 5, true),
      fill('f-0003', 19_500, 30, true)
    ])

    expect(points).toHaveLength(1)
    expect(points[0]?.id).toBe('f-0003')
    expect(points[0]?.distance_km).toBe(500)
  })
})

describe('a mis-flagged entry', () => {
  it('shifts the figures on both sides of it', () => {
    // §5.2 — "Mis-flagging full/partial shifts the figures on both sides — the
    // flag is a real field, not decoration." This test is the proof, and the
    // reason the flag is offered as a checkbox rather than inferred.
    const entries = [
      fill('f-0001', 10_000, 50, true),
      fill('f-0002', 10_400, 20, true),
      fill('f-0003', 10_900, 35, true)
    ]

    const asFlagged = consumptionPoints(entries)
    expect(asFlagged.map((point) => point.l100km)).toEqual([5_000, 7_000])

    const misflagged = [...entries]
    misflagged[1] = fill('f-0002', 10_400, 20, false)
    const after = consumptionPoints(misflagged)

    // Two intervals become one: 35 + 20 = 55 l over 900 km.
    expect(after).toHaveLength(1)
    expect(after[0]?.litres).toBe(toPump(55))
    expect(after[0]?.l100km).toBe(6_111)
  })
})

describe('the derived total of a fill-up', () => {
  it('is litres × price per litre, and matches XTRITIUM §5.1 to the kuruş', () => {
    // 29,990 l × 73,380 ₺/l → 2.200,67 ₺, the constitution's own example.
    const entry: FuelEntry = {
      id: 'f-0001',
      date: '2026-08-16',
      odometer_km: 19_764,
      litres: toPump(29.99),
      price_per_litre: toPump(73.38),
      full_tank: true,
      fuel_type: 'Kurşunsuz 95'
    }

    expect(entryTotal(entry)).toBe(220_067)
  })

  it('rounds the half-kuruş once, not the litres and then the price', () => {
    // 10 l at 8,165 ₺/l is exactly 81,65 ₺ — the price carries three decimals
    // because pump prices do, and truncating it would lose money every fill-up.
    const entry = fill('f-0001', 1_000, 10, true)
    expect(entryTotal({ ...entry, price_per_litre: toPump(8.165) })).toBe(8_165)
  })

  it('is exact across a hundred fill-ups', () => {
    // §4.3 — sums are exact because the arithmetic runs on integers. A float
    // path would drift by a kuruş or two over a year of fill-ups.
    const entry = fill('f-0001', 1_000, 33.33, true)
    const one = entryTotal({ ...entry, price_per_litre: toPump(73.38) })
    let total = 0
    for (let index = 0; index < 100; index += 1) total += one

    // 33,330 × 73,380 = 2.445.755.400, and ÷10.000 lands on 244.575,54 — the
    // half-kuruş is rounded once, here, and never again.
    expect(one).toBe(244_576)
    expect(total).toBe(24_457_600)
  })
})

describe('the precision the figure is shown at', () => {
  it('cuts three decimals down to what settings.toml asks for', () => {
    // §4.4 — decimals_consumption, default 2. The engine keeps three so the
    // setting can be raised later without the arithmetic having been discarded.
    expect(consumptionAt(6_111, 2)).toBe(611)
    expect(consumptionAt(6_115, 2)).toBe(612)
    expect(consumptionAt(6_111, 3)).toBe(6_111)
    expect(consumptionAt(6_111, 1)).toBe(61)
    expect(consumptionAt(6_111, 0)).toBe(6)
  })

  it('can be asked for more decimals than it keeps', () => {
    expect(consumptionAt(6_111, 4)).toBe(61_110)
  })
})

describe('points addressed by entry', () => {
  it('answers for the entries that have a figure, and not for the others', () => {
    const byId = consumptionById([
      fill('f-0001', 19_000, 40, true),
      fill('f-0002', 19_200, 10, false),
      fill('f-0003', 19_500, 30, true)
    ])

    expect(byId['f-0001']).toBeUndefined()
    expect(byId['f-0002']).toBeUndefined()
    expect(byId['f-0003']?.l100km).toBe(8_000)
  })
})

File: android/app/src/test/kotlin/io/github/sudomegas/tritium/storage/ConsumptionTest.kt
package io.github.sudomegas.tritium.storage

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ConsumptionTest {

    private fun full(id: String, odometer: Int, litres: Long, date: String = "2026-01-01") =
        FuelEntry(id = id, date = date, odometerKm = odometer, litres = litres, fullTank = true)

    private fun partial(id: String, odometer: Int, litres: Long, date: String = "2026-01-01") =
        FuelEntry(id = id, date = date, odometerKm = odometer, litres = litres, fullTank = false)

    @Test
    fun `the first-ever entry produces no point, full tank or not`() {
        assertEquals(emptyList<Any>(), Consumption.consumptionPoints(listOf(full("f-0001", 1000, 30000))))
        assertEquals(emptyList<Any>(), Consumption.consumptionPoints(listOf(partial("f-0001", 1000, 10000))))
    }

    @Test
    fun `a run of only partials produces no point`() {
        val entries = listOf(
            partial("f-0001", 1000, 10000),
            partial("f-0002", 1200, 12000),
            partial("f-0003", 1400, 11000),
        )
        assertEquals(emptyList<Any>(), Consumption.consumptionPoints(entries))
    }

    @Test
    fun `a partial before the first full tank is dropped, not carried forward`() {
        val entries = listOf(
            partial("f-0001", 1000, 10000),
            full("f-0002", 1300, 20000),
            full("f-0003", 1600, 25000),
        )
        val points = Consumption.consumptionPoints(entries)
        // Only one point (f-0003), and its litres must be 25000 alone — the
        // partial before f-0002 must not have leaked into this interval.
        assertEquals(1, points.size)
        assertEquals(25000L, points.single().litres)
    }

    @Test
    fun `a partial between two full tanks is summed into the LATER one`() {
        val entries = listOf(
            full("f-0001", 1000, 30000),
            partial("f-0002", 1100, 8000),
            partial("f-0003", 1150, 5000),
            full("f-0004", 1300, 20000),
        )
        val points = Consumption.consumptionPoints(entries)
        assertEquals(1, points.size)
        val point = points.single()
        assertEquals("f-0004", point.id)
        assertEquals(300, point.distanceKm)
        // 20000 + 8000 + 5000
        assertEquals(33000L, point.litres)
        // 33000 * 100 / 300 = 11000 (scaled x1000 -> 11.000 l/100km)
        assertEquals(11000L, point.l100km)
    }

    @Test
    fun `a trailing partial after the last full tank produces no point of its own`() {
        val entries = listOf(
            full("f-0001", 1000, 30000),
            full("f-0002", 1300, 20000),
            partial("f-0003", 1350, 4000),
        )
        val points = Consumption.consumptionPoints(entries)
        // Exactly one point, for f-0002 — the trailing partial never becomes
        // a point of its own, and nothing after it exists to absorb it yet.
        assertEquals(1, points.size)
        assertEquals("f-0002", points.single().id)
    }

    @Test
    fun `entries out of storage order are sorted by odometer before anything else`() {
        val entries = listOf(
            full("f-0002", 1300, 20000),
            full("f-0001", 1000, 30000),
        )
        val points = Consumption.consumptionPoints(entries)
        assertEquals(1, points.size)
        assertEquals("f-0002", points.single().id)
        assertEquals(300, points.single().distanceKm)
    }

    @Test
    fun `two entries at the same odometer produce no point, but the interval boundary still advances`() {
        val entries = listOf(
            full("f-0001", 1000, 30000),
            full("f-0002", 1300, 20000),
            // A duplicate reading right after f-0002 — same odometer.
            full("f-0003", 1300, 5000, date = "2026-01-02"),
            full("f-0004", 1600, 18000),
        )
        val points = Consumption.consumptionPoints(entries)
        // f-0002 produces a point against f-0001. f-0003 ties f-0002's
        // reading and produces none, but becomes the new interval boundary.
        // f-0004 is measured against f-0003, not f-0002.
        assertEquals(2, points.size)
        assertEquals(listOf("f-0002", "f-0004"), points.map { it.id })
        val last = points.last()
        assertEquals(300, last.distanceKm)
        assertEquals(18000L, last.litres)
    }

    @Test
    fun `mis-flagging full-tank shifts the figures on both sides of it`() {
        val correctlyFlagged = listOf(
            full("f-0001", 1000, 30000),
            partial("f-0002", 1100, 8000),
            full("f-0003", 1300, 20000),
        )
        val misFlagged = listOf(
            full("f-0001", 1000, 30000),
            // f-0002 marked full instead of partial.
            full("f-0002", 1100, 8000),
            full("f-0003", 1300, 20000),
        )

        val correctPoints = Consumption.consumptionPoints(correctlyFlagged)
        val misFlaggedPoints = Consumption.consumptionPoints(misFlagged)

        // Correct: one point at f-0003, litres = 20000 + 8000 = 28000 over 300km.
        assertEquals(1, correctPoints.size)
        assertEquals(28000L, correctPoints.single().litres)

        // Mis-flagged: TWO points now exist — f-0002 wrongly starts an
        // interval of its own — and neither matches the correct figures.
        assertEquals(2, misFlaggedPoints.size)
        assertTrue(
            "mis-flagging must change the figures on both sides, not just add a point",
            misFlaggedPoints.none { it.litres == 28000L && it.distanceKm == 300 },
        )
    }

    @Test
    fun `l100km is litres times 100 divided by distance, rounded once`() {
        val entries = listOf(
            full("f-0001", 0, 10000),
            full("f-0002", 500, 25000),
        )
        val point = Consumption.consumptionPoints(entries).single()
        // 25000 * 100 / 500 = 5000 -> 5.000 l/100km
        assertEquals(5000L, point.l100km)
    }

    @Test
    fun `consumptionAt cuts the internal 3-decimal figure to a display precision`() {
        // 5.000 l100km at 3 internal decimals -> 2 decimals -> 5.00 -> 500
        assertEquals(500L, Consumption.consumptionAt(5000L, 2))
        // -> 0 decimals -> 5
        assertEquals(5L, Consumption.consumptionAt(5000L, 0))
        // widening back to 3 decimals is a no-op
        assertEquals(5000L, Consumption.consumptionAt(5000L, 3))
    }

    @Test
    fun `consumptionById maps a point to every entry that has one, and only those`() {
        val entries = listOf(
            full("f-0001", 0, 10000),
            partial("f-0002", 200, 4000),
            full("f-0003", 500, 25000),
        )
        val byId = Consumption.consumptionById(entries)
        assertEquals(setOf("f-0003"), byId.keys)
    }

    @Test
    fun `feeding the whole entry list vs a filtered one gives different, and the whole one is correct`() {
        // The correctness rule AF4.md §1.2 states explicitly: the engine must
        // see the WHOLE list. Proved here by showing a filtered view produces
        // a wrong answer for the entry that survives the filter.
        val entries = listOf(
            full("f-0001", 1000, 30000),
            partial("f-0002", 1100, 8000),
            full("f-0003", 1300, 20000),
        )

        val whole = Consumption.consumptionById(entries)
        // A "filtered" view that drops the partial, as a range chip might if
        // applied to the engine's input instead of only to displayed rows.
        val filtered = Consumption.consumptionById(entries.filter { it.fullTank })

        assertEquals(28000L, whole.getValue("f-0003").litres)
        assertEquals(20000L, filtered.getValue("f-0003").litres) // wrong: missing the partial
        assertTrue(whole.getValue("f-0003").litres != filtered.getValue("f-0003").litres)
    }
}
